packages/frontend/src/components/Game/GameBoard/TileGrid/index.jsx

```diff
@@ -8,7 +8,11 @@ const TileGrid = ({ tiles }) => {
   const { state, dispatch } = useGameContext();
 
   const onFlipTile = (tile) => {
-    dispatch(flipTile(tile));
+    if (tile.faceUp === true || state.flipped.length === 2) {
+      return;
+    } else {
+      dispatch(flipTile(tile));
+    }
```

Comment:
Thanks for moving this into the function. Looking at it now, it might even make more sense to have it in the reducer. If the conditions are true, then it would flip the tile, otherwise it would just return the current state.

If you don't feel like moving it again, that's totally fine. I'm just thinking of it because the [Redux docs](https://redux.js.org/faq/code-structure#how-should-i-split-my-logic-between-reducers-and-action-creators-where-should-my-business-logic-go) recommend keeping as much "[business logic](https://en.wikipedia.org/wiki/Business_logic)" (i.e. logic that affects state changes) in the reducer as possible making it more maintainable because it's all in one place.